lib/Doctrine/DBAL/Driver/PDOStatement.php

```diff
@@ -264,6 +206,71 @@ public function free(): void
         parent::closeCursor();
     }
 
+    /**
+     * @param mixed $fetchArgument
+     * @param mixed $ctorArgs
+     * @param mixed ...$additionalArgs
+     */
+    private function doSetFetchMode(int $fetchMode, $fetchArgument = null, $ctorArgs = null, ...$additionalArgs): bool
+    {
+        $fetchMode = $this->convertFetchMode($fetchMode);
+
+        // This thin wrapper is necessary to shield against the weird signature
+        // of PDOStatement::setFetchMode(): even if the second and third
+        // parameters are optional, PHP will not let us remove it from this
+        // declaration.
+        try {
+            if ($fetchArgument === null && $ctorArgs === null) {
+                return parent::setFetchMode($fetchMode);
+            }
+
+            if ($ctorArgs === null) {
+                return parent::setFetchMode($fetchMode, $fetchArgument);
+            }
+
+            return parent::setFetchMode($fetchMode, $fetchArgument, $ctorArgs, ...$additionalArgs);
+        } catch (PDOException $exception) {
+            throw Exception::new($exception);
+        }
+    }
+
+    /**
+     * @param mixed $fetchArgument
+     * @param mixed $ctorArgs
+     * @param mixed ...$additionalArgs
+     *
+     * @return mixed[]
+     */
+    private function doFetchAll(
+        ?int $fetchMode = null,
+        $fetchArgument = null,
+        $ctorArgs = null,
+        ...$additionalArgs
```

Comment:
What are these additional arguments?

Reply:
They are part of the new `setFetchMode()`/`fetchAll()` signatures. I don't know if they're used inside PDO for anything yet. I'm just passing them down if they're there. Dropping them felt wrong here.

Reply:
> I don't know if they're used inside PDO for anything yet.

I don't think they do but if they add a 4th argument (which is extremely unlikely), then it may behave similar to the previous ones and would have to be filtered out. Or at least we cannot guarantee the opposite. With that said, I see to approaches:
1. Just handle the 3 arguments as they are handled now.
2. Handle all arguments in a generic way:
    ```php
    function fetch(...$args) {
        if (isset($args[0])) {
            $args[0] = convertMode($args[0]);
        }

        $slice = [];

        foreach ($args as $arg) {
            if ($arg === null) {
                break;
            }

            $slice[] = $arg.
        }

        parent::fetch(...$slice);
    }
    ````

Reply:
> I don't think they do but if they add a 4th argument (which is extremely unlikely), then it may behave similar to the previous ones and would have to be filtered out.

Filtering the currently existing parameters is only necessary for BC reasons: Doctrine's PDOStatement allows to pass `null` values where PDO's own implementation would throw errors. This does not apply to new parameters that we're about to introduce.

> Handle all arguments in a generic way:

The code you suggested breaks quite a lot of tests. I'd rather not change the logic here.

> Just handle the 3 arguments as they are handled now.

If you don't like that variadic part, we can of course just ignore it. My motivation was to stay as close to the upstream interface as possible.

Reply:
My bad, your implementation works now. I'll change it.